### PR TITLE
Speed up testing

### PR DESCRIPTION
**Pull request by f0uriest**
- Removes some redundant / unneeded tests
- Refactors some big tests to use `pytest.mark.parametrize` for more efficient parallelization
- Relabels some slow unit tests as regression tests
- Removes fixtures that require solve/optimize from as many tests as possible - replace with precomputed `desc.examples`
- Reduces resolution where possible

More can definitely be done, even parallelized unit tests still take ~30 mins, which I'd like to get down to at least half that, but that can be another PR.

### CODE REVIEW

File: tests/test_bootstrap.py [hunk 1]
```diff
@@ -1200,54 +1200,6 @@ def test_BootstrapRedlConsistency_normalization(self):
         # Results are not perfectly identical because ln(Lambda) is not quite invariant.
         np.testing.assert_allclose(results, expected, rtol=2e-3)
 
-    @pytest.mark.unit
-    @pytest.mark.solve
-    def test_BootstrapRedlConsistency_resolution(self, DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_bootstrap.py:1205]: this is redundant with `test_objective_funs.test_compute_scalar_resolution`

File: tests/test_equilibrium.py [hunk 2]
```diff
@@ -78,34 +78,6 @@ def test_compute_theta_coords(DSHAPE_current):
 
 @pytest.mark.unit
 def test_map_coordinates():
-    """Test root finding for (rho,theta,zeta) from (R,phi,Z)."""
```
Comment (f0uriest) [inline tests/test_equilibrium.py:81]: redundant with the other `test_map_coordinates`

File: tests/test_objective_funs.py [hunk 3]
```diff
@@ -626,56 +624,6 @@ def test_derivative_modes():
     np.testing.assert_allclose(H1, H3, atol=1e-10)
 
 
-@pytest.mark.unit
-@pytest.mark.xfail
-def test_rejit():
```
Comment (f0uriest) [inline tests/test_objective_funs.py:631]: changing an objective after creation is no longer recommended.

File: tests/test_objective_funs.py [hunk 4]
```diff
@@ -1360,30 +1308,6 @@ def test_boundary_error_print(capsys):
...
-def test_rebuild():
-    """Test that the objective is rebuilt correctly when needed."""
-    eq = Equilibrium(L=3, M=3)
-    f_obj = ForceBalance(eq=eq)
-    obj = ObjectiveFunction(f_obj)
-    eq.solve(maxiter=2, objective=obj)
-
-    # this would fail before v0.8.2 when trying to get objective.x
-    eq.change_resolution(L=5, M=5)
-    obj.build(eq)
```
Comment (f0uriest) [inline tests/test_objective_funs.py:1373]: build method no longer even accepts eq, this is basically passing `use_jit=eq`

File: tests/test_objective_funs.py [hunk 5]
```diff
@@ -1935,88 +1859,115 @@ def test_compute_scalar_resolution():  # noqa: C901
     np.testing.assert_allclose(f, f[-1], rtol=1e-2)
 
 
-@pytest.mark.unit
-def test_objective_no_nangrad():
```
Comment (f0uriest) [inline tests/test_objective_funs.py:1939]: this was just refactored into smaller tests to help parallelization

File: tests/test_optimizer.py [hunk 6]
```diff
@@ -559,38 +559,50 @@ def test_wrappers():
     np.testing.assert_allclose(ob.weights, con[0].weight)
 
 
-@pytest.mark.unit
-@pytest.mark.slow
-def test_all_optimizers():
```
Comment (f0uriest) [inline tests/test_optimizer.py:564]: this was just refactored into smaller tests to help parallelization

File: tests/test_optimizer.py [hunk 7]
```diff
@@ -970,60 +982,6 @@ def test_constrained_AL_scalar():
     np.testing.assert_array_less(-Dwell, ctol)
 
 
-@pytest.mark.slow
-@pytest.mark.unit
-@pytest.mark.optimize
-def test_proximal_with_PlasmaVesselDistance():
```
Comment (f0uriest) [inline tests/test_optimizer.py:976]: redundant with `test_examples.test_multiobject_optimization_prox`

File: tests/conftest.py [hunk 8]
```diff
@@ -12,45 +13,20 @@
...
+
 @pytest.fixture(scope="session")
 def TmpDir(tmpdir_factory):
     """Create a temporary directory to store testing files."""
     dir_path = tmpdir_factory.mktemp("test_results")
     return dir_path
 
 
-@pytest.fixture(scope="session")
-def SOLOVEV_vac(tmpdir_factory):
```
Comment (f0uriest) [inline tests/conftest.py:23]: this isnt really a useful tests because vacuum tokamaks have ill defined flux surfaces

File: tests/conftest.py [hunk 9]
```diff
@@ -175,24 +151,6 @@ def HELIOTRON(tmpdir_factory):
     return HELIOTRON_out
 
 
-@pytest.fixture(scope="session")
-def HELIOTRON_ex(tmpdir_factory):
```
Comment (f0uriest) [inline tests/conftest.py:179]: replaced with `desc.examples.get`

File: tests/conftest.py [hunk 10]
```diff
@@ -225,37 +183,6 @@ def HELIOTRON_vac(tmpdir_factory):
     return HELIOTRON_vacuum_out
 
 
-@pytest.fixture(scope="session")
-def HELIOTRON_vac2(tmpdir_factory):
```
Comment (f0uriest) [inline tests/conftest.py:229]: after merging #933 this will be identical to `HELIOTRON_vac`

File: tests/test_compute_funs.py [hunk 11]
```diff
@@ -1081,25 +1081,6 @@ def test_magnetic_pressure_gradient(DummyStellarator):
     )
 
 
-@pytest.mark.unit
-@pytest.mark.solve
-def test_currents(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_compute_funs.py:1086]: we now always do FSA to compute currents so this isnt testing anything

File: tests/test_examples.py [hunk 12]
```diff
@@ -50,26 +48,6 @@
 from .utils import area_difference_desc, area_difference_vmec
 
 
-@pytest.mark.regression
-@pytest.mark.solve
-def test_SOLOVEV_vacuum(SOLOVEV_vac):
```
Comment (f0uriest) [inline tests/test_examples.py:55]: removed since vacuum tokamak isnt well defined

File: tests/test_examples.py [hunk 13]
```diff
@@ -144,66 +122,6 @@ def test_HELIOTRON_vac_results(HELIOTRON_vac):
     np.testing.assert_allclose(curr(np.linspace(0, 1, 20)), 0, atol=1e-8)
 
 
-@pytest.mark.regression
-@pytest.mark.solve
-def test_HELIOTRON_vac2_results(HELIOTRON_vac, HELIOTRON_vac2):
```
Comment (f0uriest) [inline tests/test_examples.py:149]: after #933 these will be doing the same thing so no need to test

File: tests/test_examples.py [hunk 14]
```diff
@@ -144,66 +122,6 @@ def test_HELIOTRON_vac_results(HELIOTRON_vac):
...
-    iota1 = eq1.get_profile("iota")
-    iota2 = eq2.get_profile("iota")
-    np.testing.assert_allclose(curr1(np.linspace(0, 1, 20)), 0, atol=1e-8)
-    np.testing.assert_allclose(curr2(np.linspace(0, 1, 20)), 0, atol=1e-8)
-    np.testing.assert_allclose(iota1.params, iota2.params, rtol=1e-1, atol=1e-1)
-
-
-@pytest.mark.regression
-@pytest.mark.solve
-def test_force_balance_grids():
```
Comment (f0uriest) [inline tests/test_examples.py:167]: this has been a do-nothing test for a while, note the comment that the grids aren't even different anymore

File: tests/test_objective_funs.py [hunk 15]
```diff
@@ -1602,422 +1527,344 @@ def test_loss_function_asserts():
         RotationalTransform(eq=eq, loss_function=fun)
 
 
-@pytest.mark.unit
-@pytest.mark.slow
-def test_compute_scalar_resolution():  # noqa: C901
```
Comment (f0uriest) [inline tests/test_objective_funs.py:1607]: no functionality change here just broke it up into smaller pieces for parallelization and reducing code duplication

Comment (f0uriest) [reply tests/test_objective_funs.py:1607]: this will also make sure that any new added objectives get tested as well

File: tests/test_plotting.py [hunk 16]
```diff
@@ -230,40 +222,27 @@ def test_3d_B(DSHAPE_current):
 
 
 @pytest.mark.unit
-@pytest.mark.solve
-def test_3d_J(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_plotting.py:234]: basically redundant wit the other tests for 3d plotting, we already cover all the coordinate options

File: tests/test_plotting.py [hunk 17]
```diff
@@ -736,46 +690,6 @@ def test_plot_basis_fourierzernike(self):
         return fig
 
 
-class TestPlotFieldLines:
```
Comment (f0uriest) [inline tests/test_plotting.py:739]: removed with the relevant functions. Will eventually make a better version (in another PR)

File: tests/test_examples.py [hunk 18]
```diff
@@ -386,38 +304,6 @@ def test_ATF_results(tmpdir_factory):
     np.testing.assert_allclose(theta_err, 0, atol=5e-4)
 
 
-@pytest.mark.regression
-@pytest.mark.solve
-def test_ESTELL_results(tmpdir_factory):
```
Comment (f0uriest) [inline tests/test_examples.py:391]: This is basically testing the same thing as the ATF test and all of the heliotron/dshape/solovev ones, but this one takes the longest by far.

File: tests/test_objective_funs.py [hunk 19]
```diff
@@ -816,167 +929,6 @@ def test_target_profiles():
     )
 
 
-@pytest.mark.unit
-def test_plasma_vessel_distance():
```
Comment (f0uriest) [inline tests/test_objective_funs.py:820]: these were just moved to be with the others in ``TestObjectiveFunction``

File: tests/test_examples.py [hunk 20]
```diff
@@ -50,26 +48,6 @@
...
 
 
-@pytest.mark.regression
-@pytest.mark.solve
-def test_SOLOVEV_vacuum(SOLOVEV_vac):
-    """Tests that the SOLOVEV vacuum example gives no rotational transform."""
-    eq = EquilibriaFamily.load(load_from=str(SOLOVEV_vac["desc_h5_path"]))[-1]
-    data = eq.compute("|J|")
-
-    np.testing.assert_allclose(data["iota"], 0, atol=1e-16)
```
Comment (unalmis) [inline tests/test_examples.py:60]: ignoring whether it's well defined, doesn't this still unit test that those ~10 compute functions for iota yield zero as expected for this geometry?

File: tests/test_axis_limits.py [hunk 21]
```diff
@@ -340,8 +344,12 @@ def test(eq):
                 np.testing.assert_allclose(B[:, 1], B[0, 1])
                 np.testing.assert_allclose(B[:, 2], B[0, 2])
 
-        test(get("W7-X"))
-        test(get("NCSX"))
+        eq = get("W7-X")
+        eq.change_resolution(4, 4, 4, 8, 8, 8)
+        test(eq)
+        eq = get("W7-X")
```
Comment (unalmis) [inline tests/test_axis_limits.py:350]: get("NCSX")

**Review by unalmis — DISMISSED** [review]
looks okay to me besides that copying error in test axis limits

**Review by unalmis — APPROVED** [review]

File: tests/test_examples.py (hunk 20 above)
Comment (dpanici) [reply tests/test_examples.py:66]: ^ I think I agree

File: tests/test_plotting.py [hunk 22]
```diff
@@ -396,62 +375,55 @@ def test_section_J(DSHAPE_current):
 
 
 @pytest.mark.unit
-@pytest.mark.solve
 @pytest.mark.mpl_image_compare(remove_text=True, tolerance=24)
-def test_section_Z(DSHAPE_current):
+def test_section_Z():
```
Comment (dpanici) [inline tests/test_plotting.py:379]: isn't this basically redundant with the R plot, since no new code should be exercised if plotting R or Z, just a different quantiity is computed?

**Review by dpanici — CHANGES_REQUESTED** [review]
Kaya has a point with the iota test, checking that in axisymmetry the formulas for iota indeed give no rotational transform if current and nonaxisym are both zero. I'd be in favor of adding that back

File: tests/test_examples.py (hunk 20 above)
Comment (f0uriest) [reply tests/test_examples.py:66]: That functionality is already tested with `test_iota_components`

File: tests/test_compute_funs.py [hunk 23]
```diff
@@ -1166,47 +1146,6 @@ def test_boozer_transform(DSHAPE_current):
...
-@pytest.mark.unit
-def test_compute_grad_p_volume_avg():
-    """Test calculation of volume averaged pressure gradient."""
-    eq = Equilibrium()  # default pressure profile is 0 pressure
-    pres_grad_vol_avg = eq.compute("<|grad(p)|>_vol")["<|grad(p)|>_vol"]
-    np.testing.assert_allclose(pres_grad_vol_avg, 0)
-
-
-@pytest.mark.unit
-def test_compare_quantities_to_vmec():
```
Comment (f0uriest) [inline tests/test_compute_funs.py:1178]: already covered with `test_vmec_save`

File: tests/test_compute_funs.py [hunk 24]
```diff
@@ -1166,47 +1146,6 @@ def test_boozer_transform(DSHAPE_current):
     )
 
 
-@pytest.mark.unit
-def test_compute_grad_p_volume_avg():
```
Comment (f0uriest) [inline tests/test_compute_funs.py:1170]: not sure what this is really testing?

File: tests/test_equilibrium.py [hunk 25]
```diff
@@ -22,42 +21,10 @@
 
 
 @pytest.mark.unit
-@pytest.mark.solve
-def test_compute_geometry(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_equilibrium.py:26]: redundant with `test_vmec_save`

File: tests/test_equilibrium.py [hunk 26]
```diff
@@ -325,34 +261,6 @@ def test_resolution():
     assert eq1.R_basis.NFP == 5
 
 
-@pytest.mark.unit
-def test_symmetry():
```
Comment (f0uriest) [inline tests/test_equilibrium.py:329]: redundant with other `test_eq_change_symmetry`

File: tests/test_objective_funs.py [hunk 27]
```diff
@@ -711,69 +873,6 @@ def test_derivative_modes():
...
-    objFun.objectives[0].y = 2
-    np.testing.assert_allclose(objFun.compute_scaled_error(x), f)
-    np.testing.assert_allclose(objFun.jac_scaled(x), J)
-    objFun.jit()
-    np.testing.assert_allclose(objFun.compute_scaled_error(x), [-7164, 836, 828])
-    np.testing.assert_allclose(objFun.jac_scaled(x), J * 4 / 3)
-
-
-@pytest.mark.unit
-def test_generic_compute():
```
Comment (f0uriest) [inline tests/test_objective_funs.py:765]: redundant with other tests in this file

File: tests/test_plotting.py [hunk 28]
```diff
@@ -60,581 +60,489 @@ def test_kwarg_warning(DummyStellarator):
...
-    grid = LinearGrid(N=20, NFP=eq.NFP)
-    fig, ax, data = plot_1d(
-        eq, "a_major/a_minor", grid=grid, figsize=(4, 4), return_data=True
-    )
-    assert "a_major/a_minor" in data.keys()
-    return fig
-
-
-@pytest.mark.unit
-def test_1d_fsa_consistency():
```
Comment (f0uriest) [inline tests/test_plotting.py:92]: Redundant with `plot_1d_iota_radial` for fixed current equilibrium

File: tests/test_plotting.py [hunk 29]
```diff
@@ -60,581 +60,489 @@ def test_kwarg_warning(DummyStellarator):
...
 
-    return fig
 
+class TestPlot2D:
+    """Tests for plot_2d."""
 
-@pytest.mark.unit
-@pytest.mark.solve
-@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_1d)
-def test_fsa_G(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_plotting.py:363]: redundant with `test_fsa_I`

File: tests/test_plotting.py [hunk 30]
```diff
@@ -60,581 +60,489 @@ def test_kwarg_warning(DummyStellarator):
...
-    lg = LinearGrid(rho=np.linspace(0, 1, 30))
-    test("magnetic well")
-    test("magnetic well", grid=lg)
-    test("current", with_sqrt_g=False)
-    test("current", with_sqrt_g=False, grid=lg)
-
-
-@pytest.mark.unit
-@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_1d)
-def test_1d_dpdr():
```
Comment (f0uriest) [inline tests/test_plotting.py:118]: redundant with `test_1d_p`

File: tests/test_plotting.py [hunk 31]
```diff
@@ -60,581 +60,489 @@ def test_kwarg_warning(DummyStellarator):
...
-    assert "zeta" in data.keys()
-    assert "sqrt(g)" in data.keys()
-
-    return fig
-
-
-@pytest.mark.unit
-@pytest.mark.solve
-@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_2d)
-def test_2d_lambda(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_plotting.py:209]: redundant with other plot_2d tests

File: tests/test_plotting.py [hunk 32]
```diff
@@ -60,581 +60,489 @@ def test_kwarg_warning(DummyStellarator):
...
+        for string in ["e^rho", "normalization", "theta", "zeta"]:
+            assert string in data.keys()
+        assert data["normalization"] == 1
 
+        return fig
 
-@pytest.mark.unit
-@pytest.mark.solve
-@pytest.mark.mpl_image_compare(remove_text=True, tolerance=24)
-def test_section_Z(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_plotting.py:401]: redundant with `plot_section_F`

File: tests/test_plotting.py [hunk 33]
```diff
@@ -60,581 +60,489 @@ def test_kwarg_warning(DummyStellarator):
...
+            eq, "|F|", norm_F=True, figsize=(4, 4), return_data=True, grid=grid
+        )
+        for string in ["|F|", "theta", "zeta"]:
+            assert string in data.keys()
+        return fig
 
-@pytest.mark.unit
-@pytest.mark.solve
-@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_2d)
-def test_section_R(DSHAPE_current):
```
Comment (f0uriest) [inline tests/test_plotting.py:411]: redundant with `plot_section_F`

File: tests/test_plotting.py
Comment (f0uriest) [inline tests/test_plotting.py:1]: git struggles with the diff here. Mostly was just re-organizing things into classes for testing each of the different functions. I left specific comments on the tests that were  actually removed.

**Review by dpanici — APPROVED** [review]

File: tests/test_plotting.py (hunk 28 above)
Comment (unalmis) [reply tests/test_plotting.py:92]: This test was added in #376 for issue #368. Since the git diff is ugly, for convenience here is the test on master: https://github.com/PlasmaControl/DESC/blob/5801ea8238a45b6f6674cc0d2ebe67d9d598242f/tests/test_plotting.py#L92

 `test_1d_iota_radial` tests whether the plot of the rotational transform of DSHAPE_current changes. It fails to test whether the plot is accurate in the first place. Prior to #376, `iota` would be computed (inaccurately) on a 1d grid, which that pull request corrects so that it is computed on a 2d grid, yet the baseline image for `iota` does not change for `test_1d_iota_radial`.

The correction is measurable with quantities computed in `test_1d_fsa_consistency`, i.e. magnetic well as reported in the issue.

If the purpose of `test_1d_iota_radial` is solely to test whether an 2d grid is used for flux surface average quantities, we should remove `test_1d_iota_radial` and keep `test_1d_fsa_consistency`. If the rotational transform image is an important test in particular, we should keep both tests, or merge them.

File: tests/test_plotting.py (hunk 29 above)
Comment (unalmis) [reply tests/test_plotting.py:363]: If anyone else was initially uncomfortable with removing these tests, I'm okay with it since I think it's redundant with the two tests (`test_fsa_I` and `test_compute_everything`).

File: tests/test_constrain_current.py [hunk 34]
```diff
@@ -192,6 +190,8 @@ def test(stellarator, grid_type):
                 )
 
         # Only makes sense to test on configurations with fixed current profiles.
-        test(HELIOTRON_vac, QuadratureGrid)
-        test(HELIOTRON_vac, ConcentricGrid)
-        test(HELIOTRON_vac, LinearGrid)
+        eq = desc.examples.get("ESTELL")
```
Comment (unalmis) [inline tests/test_constrain_current.py:193]: When testing with heliotron_vac, we needed to add atol=1e-7 to this test (line 187) because rtol -> infinity when the expected current was zero. If we're replacing the live solve test with a saved file, so time to solve is not an issue, should we use a less trivial example for the test (i.e. nonzero current)?

**Review by unalmis — CHANGES_REQUESTED** [review]
I think of the two redundant 1d plotting tests, the wrong one was removed

File: tests/test_plotting.py (hunk 28 above)
Comment (f0uriest) [reply tests/test_plotting.py:92]: I think they're both somewhat redundant, since the logic in `eq.compute` now ensures that the correct grid is used for computing FSA quantities (and there are a number of other tests that check that, eg `test_compute_everything`). I kept `iota_radial` just to exercise that bit of the code, and its faster than `1d_fsa_consistency`

Comment (unalmis) [reply tests/test_plotting.py:92]: Ok but we no longer have a test that confirms whether the plotting is done on the correct grid. Because test_1d_iota_radial will pass if the computation is done on a 1d grid too, and test_compute_everything doesn't check how we compute what we plot.

**Review by unalmis — APPROVED** [review]

File: tests/test_plotting.py (hunk 28 above)
Comment (f0uriest) [reply tests/test_plotting.py:92]: I think it should be fine, since `plotting._compute` just calls `eq.compute` which includes logic for always using the correct grid.

File: tests/test_constrain_current.py (hunk 34 above)
Comment (f0uriest) [reply tests/test_constrain_current.py:193]: I think it's fine. `atol=1e-7` is about as strict as you can get for the current profile because `eps/mu_0 ~ 1e-8`

**Review by unalmis — APPROVED** [review]

**Review by dpanici — APPROVED** [review]
